src/System.Runtime/tests/System/Int32Tests.cs

```diff
@@ -150,131 +150,241 @@ public static void ToString_InvalidFormat_ThrowsFormatException()
         }
 
         [Theory]
         [MemberData(nameof(Parse_Valid_TestData))]
         public static void Parse(string value, NumberStyles style, IFormatProvider provider, int expected)
         {
+            bool isDefaultProvider = provider == null || provider == new NumberFormatInfo();
             int result;
-            // If no style is specified, use the (String) or (String, IFormatProvider) overload
-            if (style == NumberStyles.Integer)
+            if ((style & ~NumberStyles.Integer) == 0 && style != NumberStyles.None)
             {
-                Assert.True(int.TryParse(value, out result));
-                Assert.Equal(expected, result);
-
-                Assert.Equal(expected, int.Parse(value));
-
-                // If a format provider is specified, but the style is the default, use the (String, IFormatProvider) overload
-                if (provider != null)
+                // Use Parse(string) or Parse(string, IFormatProvider)
+                if (isDefaultProvider)
                 {
-                    Assert.Equal(expected, int.Parse(value, provider));
+                    Assert.True(int.TryParse(value, out result));
+                    Assert.Equal(expected, result);
+
+                    Assert.Equal(expected, int.Parse(value));
                 }
-            }
 
-            // If a format provider isn't specified, test the default one, using a new instance of NumberFormatInfo
-            Assert.True(int.TryParse(value, style, provider ?? new NumberFormatInfo(), out result));
+                Assert.Equal(expected, int.Parse(value, provider));
+            }
+            
+            // Use Parse(string, NumberStyles, IFormatProvider)
+            Assert.True(int.TryParse(value, style, provider, out result));
             Assert.Equal(expected, result);
 
-            // If a format provider isn't specified, test the default one, using the (String, NumberStyles) overload
-            if (provider == null)
+            Assert.Equal(expected, int.Parse(value, style, provider));
+
+            if (isDefaultProvider)
             {
+                // Use Parse(string, NumberStyles) or Parse(string, NumberStyles, IFormatProvider)
+                Assert.True(int.TryParse(value, style, new NumberFormatInfo(), out result));
+                Assert.Equal(expected, result);
+                
                 Assert.Equal(expected, int.Parse(value, style));
+                Assert.Equal(expected, int.Parse(value, style, new NumberFormatInfo()));
             }
-            Assert.Equal(expected, int.Parse(value, style, provider ?? new NumberFormatInfo()));
         }
 
         public static IEnumerable<object[]> Parse_Invalid_TestData()
         {
             NumberStyles defaultStyle = NumberStyles.Integer;
-
-            NumberFormatInfo customFormat = new NumberFormatInfo();
-            customFormat.CurrencySymbol = "$";
-            customFormat.NumberDecimalSeparator = ".";
-
+            
```

Comment:
Good idea. What I did was generalised some tests where I could (e..g AllowLeadingSign), and then tested the rest with NumberStyles.Integer and NumberStyles.Any